lib/action.ml

```diff
@@ -9,7 +9,59 @@ exception Action_error of string
 let action_error msg = raise (Action_error msg)
 let log = Log.from "action"
 
+module SlackUsername (Slack_api : Api.Slack) = struct
+  let email_canonical_regex = Re2.create_exn "\\+[^]]*|@[^]]*" (* Match email domain and everything after '+' *)
+
+  let username_ignored_chars = [ '.'; '-' ]
+  let username_to_slack_id_tbl = Stringtbl.empty ()
+
+  let canonicalize_username username =
+    username
+    |> String.to_seq
+    |> Seq.filter (Fun.negate @@ flip List.mem username_ignored_chars)
+    |> String.of_seq
+    |> String.lowercase_ascii
+
+  let canonicalize_email_username email =
+    email |> Re2.replace_exn ~f:(fun _ -> "") email_canonical_regex |> canonicalize_username
+
+  let refresh_username_to_slack_id_tbl ~ctx =
+    log#info "now updating slack username mapping";
+    match%lwt Slack_api.list_users ~ctx () with
+    | Error e ->
+      log#warn "couldn't fetch list of Slack users: %s" e;
+      Lwt.return_unit
+    | Ok res ->
+      List.iter
+        (fun (user : Slack_t.user) ->
+          match user.profile.email with
+          | None -> ()
+          | Some email ->
+            let username = canonicalize_email_username email in
+            Stringtbl.replace username_to_slack_id_tbl username user.id
+        )
+        res.members;
+      Lwt.return_unit
+
+  let rec refresh_username_to_slack_id_tbl_background_lwt ~ctx : unit Lwt.t =
+    let%lwt () = refresh_username_to_slack_id_tbl ~ctx in
+    let%lwt () = Lwt_unix.sleep 120. in
+    (* Updates mapping every 2 minutes *)
```

Comment:
2 minute interval is a little short? I imagine this mapping won't be updated very often, so maybe updating it every day (and/or on program startup) is enough.